src/Compilers/Core/Portable/Collections/TopologicalSort.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A helper class that contains a topological sort algorithm.
+    /// </summary>
+    internal static class TopologicalSort
+    {
+        /// <summary>
+        /// Produce a topological sort of a given directed acyclic graph, given a set of nodes which include all nodes
+        /// that have no predecessors. Any nodes not in the given set, but reachable through successors, will be added
+        /// to the result. This is an iterative rather than recursive implementation, so it is unlikely to cause a stack
+        /// overflow.
+        /// </summary>
+        /// <typeparam name="TNode">The type of the node</typeparam>
+        /// <param name="nodes">Any subset of the nodes that includes all nodes with no predecessors</param>
+        /// <param name="successors">A function mapping a node to its set of successors</param>
+        /// <returns>A list of all reachable nodes, in which each node always precedes its successors</returns>
+        public static ImmutableArray<TNode> IterativeSort<TNode>(IEnumerable<TNode> nodes, Func<TNode, IEnumerable<TNode>> successors)
```

Comment:
Perhaps want an overload with ImmutableArrays as inputs, but waiting until the actual use also seems ok.